cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,682 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "arrow_filter_policy.cuh"
+#include "compact_protocol_reader.hpp"
+#include "io/parquet/parquet.hpp"
+#include "reader_impl_helpers.hpp"
+
+#include <cudf/ast/detail/expression_transformer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/transform.hpp>
+#include <cudf/detail/utilities/logger.hpp>
+#include <cudf/hashing/detail/xxhash_64.cuh>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_vector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuco/bloom_filter_ref.cuh>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <future>
+#include <numeric>
+#include <optional>
+
+namespace cudf::io::parquet::detail {
+namespace {
+
+/**
+ * @brief Converts bloom filter membership results (for each column chunk) to a device column.
+ *
+ */
+struct bloom_filter_caster {
+  cudf::device_span<void* const> buffer_ptrs;
+  cudf::device_span<size_t const> buffer_sizes;
+  host_span<Type const> parquet_types;
+  size_t num_row_groups;
+  size_t num_equality_columns;
+
+  template <typename T, bool timestamp_is_int96 = false>
+  std::unique_ptr<cudf::column> query_bloom_filter(cudf::size_type equality_col_idx,
+                                                   cudf::data_type dtype,
+                                                   ast::literal* const& literal,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using key_type    = T;
+    using hasher_type = cudf::hashing::detail::XXHash_64<key_type>;
+    using policy_type = cuco::arrow_filter_policy<key_type, hasher_type>;
+    using word_type   = typename policy_type::word_type;
+
+    // Check if the literal has the same type as the column
+    CUDF_EXPECTS(dtype.id() == literal->get_data_type().id(),
```

Comment:
Can we use a form of `cudf::have_same_types(column_view const& lhs, scalar const& rhs)` from `<cudf/utilities/type_checks.hpp>`? Raw type checks are not good practice, I spent a lot of time eradicating bugs of this flavor: https://github.com/rapidsai/cudf/blob/branch-25.02/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#comparing-data-types

It may require you to plumb through the column itself as a parameter and not just the `data_type`.

Reply:
Thank you for pointing this out. Let me think of a fix for it.

Reply:
CC @vuule